diff --git a/amd/hipcc/src/hipBin_amd.h b/amd/hipcc/src/hipBin_amd.h
index e845139080c4..eedd52bbd7e6 100644
--- a/amd/hipcc/src/hipBin_amd.h
+++ b/amd/hipcc/src/hipBin_amd.h
@@ -201,16 +201,11 @@ void HipBinAmd::constructCompilerPath() {
   const EnvVariables& envVariables = getEnvVariables();
   if (envVariables.hipClangPathEnv_.empty()) {
     fs::path hipClangPath;
-    if (isWindows()) {
-      compilerPath = getHipPath();
-      hipClangPath = compilerPath;
-      hipClangPath /= "bin";
-    } else {
-      compilerPath = getRoccmPath();
-      hipClangPath = compilerPath;
-      hipClangPath /= "lib/llvm/bin";
-    }
-
+    compilerPath = getRoccmPath();
+    hipClangPath = compilerPath;
+    hipClangPath /= "lib";
+    hipClangPath /= "llvm";
+    hipClangPath /= "bin";
     compilerPath = hipClangPath.string();
   } else {
     compilerPath = envVariables.hipClangPathEnv_;
@@ -226,32 +221,17 @@ const string& HipBinAmd::getCompilerPath() const {
 void HipBinAmd::printCompilerInfo() const {
   const string& hipClangPath = getCompilerPath();
   const string& hipPath = getHipPath();
-  if (isWindows()) {
-    string cmd = hipClangPath + "/clang++ --version";
-    system(cmd.c_str());  // hipclang version
-    cout << "llc-version :" << endl;
-    cmd = hipClangPath + "/llc --version";
-    system(cmd.c_str());  // llc version
-    cout << "hip-clang-cxxflags :" << endl;
-    cmd = hipPath + "/bin/hipcc  --cxxflags";
-    system(cmd.c_str());  // cxx flags
-    cout << endl << "hip-clang-ldflags :" << endl;
-    cmd = hipPath + "/bin/hipcc --ldflags";
-    system(cmd.c_str());  // ld flags
-    cout << endl;
-  } else {
-    string cmd = hipClangPath + "/clang++ --version";
-    system(cmd.c_str());  // hipclang version
-    cmd = hipClangPath + "/llc --version";
-    system(cmd.c_str());  // llc version
-    cout << "hip-clang-cxxflags :" << endl;
-    cmd = hipPath + "/bin/hipcc --cxxflags";
-    system(cmd.c_str());  // cxx flags
-    cout << endl << "hip-clang-ldflags :" << endl;
-    cmd = hipPath + "/bin/hipcc --ldflags";
-    system(cmd.c_str());  // ldflags version
-    cout << endl;
-  }
+  string cmd = hipClangPath + "/clang++ --version";
+  system(cmd.c_str());  // hipclang version
+  cmd = hipClangPath + "/llc --version";
+  system(cmd.c_str());  // llc version
+  cout << "hip-clang-cxxflags :" << endl;
+  cmd = hipPath + "/bin/hipcc --cxxflags";
+  system(cmd.c_str());  // cxx flags
+  cout << endl << "hip-clang-ldflags :" << endl;
+  cmd = hipPath + "/bin/hipcc --ldflags";
+  system(cmd.c_str());  // ldflags version
+  cout << endl;
 }
 
 string HipBinAmd::getCompilerVersion() {
@@ -293,17 +273,19 @@ string HipBinAmd::getCppConfig() {
   hipPathInclude = hipPath;
   hipPathInclude /= "include";
   if (isWindows()) {
+    // -I{hipPathInclude}/
     cppConfig += " -I" + hipPathInclude.string();
     cppConfigFs = cppConfig;
     cppConfigFs /= "/";
   } else {
+    // -I{hipPathInclude} -I{hsaPath}/include
     const string& hsaPath = getHsaPath();
     cppConfig += " -I" + hipPathInclude.string() +
                  " -I" + hsaPath;
     cppConfigFs = cppConfig;
     cppConfigFs /= "include";
-    cppConfig = cppConfigFs.string();
   }
+  cppConfig = cppConfigFs.string();
   return cppConfig;
 }
 
@@ -864,7 +846,9 @@ void HipBinAmd::executeHipCCCmd(vector<string> argv) {
 
   // to avoid using dk linker or MSVC linker
   if (isWindows()) {
-    HIPLDFLAGS += " -fuse-ld=lld --ld-path=\"" + hipClangPath + "/lld-link.exe\"";
+    fs::path ldPath = hipClangPath;
+    ldPath /= "lld-link.exe";
+    HIPLDFLAGS += " -fuse-ld=lld --ld-path=\"" + ldPath.string() + "\"";
   }
 
   if (!compileOnly) {
diff --git a/amd/hipcc/src/hipBin_base.h b/amd/hipcc/src/hipBin_base.h
index ea37e6fd12fc..4aa7431fdba6 100644
--- a/amd/hipcc/src/hipBin_base.h
+++ b/amd/hipcc/src/hipBin_base.h
@@ -320,16 +320,34 @@ void HipBinBase::readEnvVariables() {
 
 // constructs the HIP path
 void HipBinBase::constructHipPath() {
-  // we need to use --hip-path option
+  // The --hip-path argument option takes precedence over all other settings.
   string hip_path_name = gethip_pathOption();
   if (!hip_path_name.empty()) {
     variables_.hipPathEnv_ = hip_path_name;
-  } else if (envVariables_.hipPathEnv_.empty()) {
-    fs::path full_path(hipcc::utils::getSelfPath());
-    variables_.hipPathEnv_ = (full_path.parent_path()).string();
-  } else {
+    return;
+  }
+
+  fs::path full_path(hipcc::utils::getSelfPath());
+  fs::path parent_path = full_path.parent_path();
+
+  // Next, check for `../lib/llvm/bin/`, the standard ROCm install structure.
+  fs::path llvm_path = parent_path / "lib" / "llvm" / "bin";
+  if (fs::exists(llvm_path)) {
+    variables_.hipPathEnv_ = parent_path.string();
+    return;
+  }
+
+  // Otherwise, check the HIP_PATH environment variable from the HIP SDK.
+  // Normally an environment variable setting could take precedence over an
+  // implicit path, but this environment variable is set by system-wide installs
+  // and self-contained builds/installs should not be reading that global state.
+  if (!envVariables_.hipPathEnv_.empty()) {
     variables_.hipPathEnv_ = envVariables_.hipPathEnv_;
+    return;
   }
+
+  // Finally, fallback to the parent path (the standard ROCm install structure).
+  variables_.hipPathEnv_ = parent_path.string();
 }
 
 
diff --git a/clang/lib/Driver/ToolChains/AMDGPU.cpp b/clang/lib/Driver/ToolChains/AMDGPU.cpp
index 300c72b6a508..1c4aef089a01 100644
--- a/clang/lib/Driver/ToolChains/AMDGPU.cpp
+++ b/clang/lib/Driver/ToolChains/AMDGPU.cpp
@@ -242,8 +242,17 @@ RocmInstallationDetector::getInstallationPathCandidates() {
 
     // Some versions of the rocm llvm package install to /opt/rocm/llvm/bin
     // Some versions of the aomp package install to /opt/rocm/aomp/bin
-    if (ParentName == "llvm" || ParentName.starts_with("aomp"))
+    if (ParentName == "llvm" || ParentName.starts_with("aomp")) {
       ParentDir = llvm::sys::path::parent_path(ParentDir);
+      ParentName = llvm::sys::path::filename(ParentDir);
+
+      // Some versions of the rocm llvm package install to 
+      // /opt/rocm/lib/llvm/bin, so also back up if within the lib dir still.
+      if (ParentName == "lib") {
+        ParentDir = llvm::sys::path::parent_path(ParentDir);
+      }
+    }
+
     // Some versions of the aomp package install to /opt/rocm/aomp/bin
     // and it seems ParentDir is already pointing to correct place.
     return Candidate(ParentDir.str(), /*StrictChecking=*/true);
