diff --git a/tools/time-trace/rccl-TimeTrace.sh b/tools/time-trace/rccl-TimeTrace.sh
index 7f0dee6f..765b40b3 100755
--- a/tools/time-trace/rccl-TimeTrace.sh
+++ b/tools/time-trace/rccl-TimeTrace.sh
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 # Directory path to search for JSON files
-directory="../../build/release"
+directory=${build_dir:-"../../build/release"}
 
 if command -v pip &>/dev/null; then
     echo "pip is already installed."
@@ -47,4 +47,4 @@ mv temp_file.txt "$directory/.ninja_log"
 mv "$directory/.ninja_log" "$directory/time_trace.log"
 
 # Run the python program
-python3 time_trace_generator.py --min_val 5 --include_linking
\ No newline at end of file
+python3 time_trace_generator.py --min_val 5 --include_linking --log_file_path $directory/time_trace.log
