diff --git a/Tensile/AsmCaps.py b/Tensile/AsmCaps.py
index 26f6a98b..643c3d7f 100644
--- a/Tensile/AsmCaps.py
+++ b/Tensile/AsmCaps.py
@@ -741,6 +741,50 @@ def getCapabilitiesCache(rocmVersion: NamedTuple) -> dict:
                   'v_mov_b64': False,
                   'v_pk_fma_f16': True,
                   'v_pk_fmac_f16': False},
+   (11, 0, 3): {'HasAddLshl': True,
+                  'HasAtomicAdd': True,
+                  'HasDirectToLdsDest': False,
+                  'HasDirectToLdsNoDest': False,
+                  'HasExplicitCO': True,
+                  'HasExplicitNC': True,
+                  'HasGLCModifier': True,
+                  'HasNTModifier': False,
+                  'HasLshlOr': True,
+                  'HasMFMA': False,
+                  'HasMFMA_b8': False,
+                  'HasMFMA_bf16_1k': False,
+                  'HasMFMA_bf16_original': False,
+                  'HasMFMA_constSrc': False,
+                  'HasMFMA_f64': False,
+                  'HasMFMA_f8': False,
+                  'HasMFMA_i8_908': False,
+                  'HasMFMA_i8_940': False,
+                  'HasMFMA_vgpr': False,
+                  'HasMFMA_xf32': False,
+                  'HasSMulHi': True,
+                  'HasWMMA': True,
+                  'KernargPreloading': False,
+                  'MaxLgkmcnt': 15,
+                  'MaxVmcnt': 63,
+                  'SupportedISA': True,
+                  'SupportedSource': True,
+                  'VOP3v_dot4_i32_i8': False,
+                  'v_dot2_f32_f16': True,
+                  'v_dot2c_f32_f16': True,
+                  'v_dot4_i32_i8': False,
+                  'v_dot4c_i32_i8': False,
+                  'v_fma_f16': True,
+                  'v_fma_f32': True,
+                  'v_fma_f64': True,
+                  'v_fma_mix_f32': True,
+                  'v_fmac_f16': False,
+                  'v_fmac_f32': True,
+                  'v_mac_f16': False,
+                  'v_mac_f32': False,
+                  'v_mad_mix_f32': False,
+                  'v_mov_b64': False,
+                  'v_pk_fma_f16': True,
+                  'v_pk_fmac_f16': False},
      (11, 5, 0): {'HasAddLshl': True,
                   'HasAtomicAdd': True,
                   'HasDirectToLdsDest': False,
diff --git a/Tensile/Common.py b/Tensile/Common.py
index 9537b770..3f82448b 100644
--- a/Tensile/Common.py
+++ b/Tensile/Common.py
@@ -249,7 +249,7 @@ globalParameters["SupportedISA"] = [(8,0,3),
                                     (9,0,0), (9,0,6), (9,0,8), (9,0,10),
                                     (9,4,2), (9,5,0),
                                     (10,1,0), (10,1,1), (10,1,2), (10,3,0), (10,3,1),
-                                    (11,0,0), (11,0,1), (11,0,2),
+                                    (11,0,0), (11,0,1), (11,0,2), (11,0,3),
                                     (11,5,0), (11,5,1),
                                     (12,0,0), (12,0,1)] # assembly kernels writer supports these architectures
 
@@ -308,7 +308,7 @@ globalParameters["SeparateArchitectures"] = False # write Tensile library metada
 
 globalParameters["LazyLibraryLoading"] = False # Load library and code object files when needed instead of at startup
 
-globalParameters["IgnoreAsmCapCache"] = False # Ignore checking for discrepancies between derived and cached asm caps
+globalParameters["IgnoreAsmCapCache"] = True # Ignore checking for discrepancies between derived and cached asm caps
 
 globalParameters["ExperimentalLogicDir"] = "/experimental/"
 
@@ -325,7 +325,7 @@ architectureMap = {
   'gfx950':'gfx950', 'gfx950:xnack+':'gfx950', 'gfx950:xnack-':'gfx950',
   'gfx1010':'navi10', 'gfx1011':'navi12', 'gfx1012':'navi14',
   'gfx1030':'navi21', 'gfx1031':'navi22', 'gfx1032':'navi23', 'gfx1034':'navi24', 'gfx1035':'rembrandt',
-  'gfx1100':'navi31', 'gfx1101':'navi32', 'gfx1102':'navi33',
+  'gfx1100':'navi31', 'gfx1101':'navi32', 'gfx1102':'navi33', 'gfx1103':'phoenix',
   'gfx1150':'strixpoint', 'gfx1151':'strixhalo',
   'gfx1200':'gfx1200',
   'gfx1201':'gfx1201'
diff --git a/Tensile/Source/CMakeLists.txt b/Tensile/Source/CMakeLists.txt
index 8279cfb0..7735220a 100644
--- a/Tensile/Source/CMakeLists.txt
+++ b/Tensile/Source/CMakeLists.txt
@@ -51,9 +51,9 @@ if(CMAKE_CXX_COMPILER MATCHES ".*/hipcc$" OR CMAKE_CXX_COMPILER MATCHES ".*clang
 endif()
 
 if(CMAKE_CXX_COMPILER STREQUAL "hipcc")
-  set(TENSILE_GPU_ARCHS gfx803 gfx900 gfx906:xnack- gfx908:xnack- gfx90a:xnack- gfx1010 gfx1011 gfx1012 gfx1030 gfx1031 gfx1032 gfx1034 gfx1035 gfx1100 gfx1101 gfx1102 gfx1150 gfx1151 CACHE STRING "GPU architectures")
+  set(TENSILE_GPU_ARCHS gfx803 gfx900 gfx906:xnack- gfx908:xnack- gfx90a:xnack- gfx1010 gfx1011 gfx1012 gfx1030 gfx1031 gfx1032 gfx1034 gfx1035 gfx1100 gfx1101 gfx1102 gfx1103 gfx1150 gfx1151 CACHE STRING "GPU architectures")
 else()
-  set(TENSILE_GPU_ARCHS gfx803 gfx900 gfx906 gfx908 gfx90a gfx1010 gfx1011 gfx1012 gfx1030 gfx1031 gfx1032 gfx1034 gfx1035 gfx1100 gfx1101 gfx1102 gfx1150 gfx1151 CACHE STRING "GPU architectures")
+  set(TENSILE_GPU_ARCHS gfx803 gfx900 gfx906 gfx908 gfx90a gfx1010 gfx1011 gfx1012 gfx1030 gfx1031 gfx1032 gfx1034 gfx1035 gfx1100 gfx1101 gfx1102 gfx1103 gfx1150 gfx1151 CACHE STRING "GPU architectures")
 endif()
 
 include(CMakeDependentOption)
diff --git a/Tensile/Source/lib/include/Tensile/AMDGPU.hpp b/Tensile/Source/lib/include/Tensile/AMDGPU.hpp
index 0a1f9e67..19930a3c 100644
--- a/Tensile/Source/lib/include/Tensile/AMDGPU.hpp
+++ b/Tensile/Source/lib/include/Tensile/AMDGPU.hpp
@@ -60,6 +60,7 @@ namespace Tensile
             gfx906  = 906,
             gfx908  = 908,
             gfx90a  = 910,
+            gfx90c  = 912,
             gfx942  = 942,
             gfx950  = 950,
             gfx1010 = 1010,
@@ -73,6 +74,7 @@ namespace Tensile
             gfx1100 = 1100,
             gfx1101 = 1101,
             gfx1102 = 1102,
+            gfx1103 = 1103,
             gfx1150 = 1150,
             gfx1151 = 1151
         };
@@ -91,6 +93,8 @@ namespace Tensile
                 return "gfx908";
             case AMDGPU::Processor::gfx90a:
                 return "gfx90a";
+            case AMDGPU::Processor::gfx90c:
+                return "gfx90c";
             case AMDGPU::Processor::gfx942:
                 return "gfx942";
             case AMDGPU::Processor::gfx950:
@@ -117,6 +121,8 @@ namespace Tensile
                 return "gfx1101";
             case AMDGPU::Processor::gfx1102:
                 return "gfx1102";
+            case AMDGPU::Processor::gfx1103:
+                return "gfx1103";
             case AMDGPU::Processor::gfx1150:
                 return "gfx1150";
             case AMDGPU::Processor::gfx1151:
@@ -147,6 +153,10 @@ namespace Tensile
             {
                 return AMDGPU::Processor::gfx90a;
             }
+            else if(deviceString.find("gfx90c") != std::string::npos)
+            {
+                return AMDGPU::Processor::gfx90c;
+            }
             else if(deviceString.find("gfx942") != std::string::npos)
             {
                 return AMDGPU::Processor::gfx942;
@@ -183,6 +193,10 @@ namespace Tensile
             {
                 return AMDGPU::Processor::gfx1102;
             }
+            else if(deviceString.find("gfx1103") != std::string::npos)
+            {
+                return AMDGPU::Processor::gfx1103;
+            }
             else if(deviceString.find("gfx1150") != std::string::npos)
             {
                 return AMDGPU::Processor::gfx1150;
diff --git a/Tensile/Source/lib/include/Tensile/PlaceholderLibrary.hpp b/Tensile/Source/lib/include/Tensile/PlaceholderLibrary.hpp
index a7687315..4f0107d6 100644
--- a/Tensile/Source/lib/include/Tensile/PlaceholderLibrary.hpp
+++ b/Tensile/Source/lib/include/Tensile/PlaceholderLibrary.hpp
@@ -44,6 +44,7 @@ namespace Tensile
         gfx906,
         gfx908,
         gfx90a,
+        gfx90c,
         gfx942,
         gfx950,
         gfx1010,
@@ -57,6 +58,7 @@ namespace Tensile
         gfx1100,
         gfx1101,
         gfx1102,
+        gfx1103,
         gfx1150,
         gfx1151,
         All
@@ -79,6 +81,8 @@ namespace Tensile
             return "TensileLibrary_*_gfx908";
         case LazyLoadingInit::gfx90a:
             return "TensileLibrary_*_gfx90a";
+        case LazyLoadingInit::gfx90c:
+            return "TensileLibrary_*_gfx90c";
         case LazyLoadingInit::gfx942:
             return "TensileLibrary_*_gfx942";
         case LazyLoadingInit::gfx950:
@@ -105,6 +109,8 @@ namespace Tensile
             return "TensileLibrary_*_gfx1101";
         case LazyLoadingInit::gfx1102:
             return "TensileLibrary_*_gfx1102";
+        case LazyLoadingInit::gfx1103:
+            return "TensileLibrary_*_gfx1103";
         case LazyLoadingInit::gfx1150:
             return "TensileLibrary_*_gfx1150";
         case LazyLoadingInit::gfx1151:
diff --git a/Tensile/Source/lib/include/Tensile/Serialization/Predicates.hpp b/Tensile/Source/lib/include/Tensile/Serialization/Predicates.hpp
index 12d81fc9..c2c32171 100644
--- a/Tensile/Source/lib/include/Tensile/Serialization/Predicates.hpp
+++ b/Tensile/Source/lib/include/Tensile/Serialization/Predicates.hpp
@@ -220,6 +220,7 @@ namespace Tensile
                 iot::enumCase(io, value, "gfx906", AMDGPU::Processor::gfx906);
                 iot::enumCase(io, value, "gfx908", AMDGPU::Processor::gfx908);
                 iot::enumCase(io, value, "gfx90a", AMDGPU::Processor::gfx90a);
+                iot::enumCase(io, value, "gfx90c", AMDGPU::Processor::gfx90c);
                 iot::enumCase(io, value, "gfx942", AMDGPU::Processor::gfx942);
                 iot::enumCase(io, value, "gfx950", AMDGPU::Processor::gfx950);
                 iot::enumCase(io, value, "gfx1010", AMDGPU::Processor::gfx1010);
@@ -233,6 +234,7 @@ namespace Tensile
                 iot::enumCase(io, value, "gfx1100", AMDGPU::Processor::gfx1100);
                 iot::enumCase(io, value, "gfx1101", AMDGPU::Processor::gfx1101);
                 iot::enumCase(io, value, "gfx1102", AMDGPU::Processor::gfx1102);
+                iot::enumCase(io, value, "gfx1103", AMDGPU::Processor::gfx1103);
                 iot::enumCase(io, value, "gfx1150", AMDGPU::Processor::gfx1150);
                 iot::enumCase(io, value, "gfx1151", AMDGPU::Processor::gfx1151);
             }
diff --git a/Tensile/TensileCreateLib/ParseArguments.py b/Tensile/TensileCreateLib/ParseArguments.py
index c95731ba..a41feb7f 100644
--- a/Tensile/TensileCreateLib/ParseArguments.py
+++ b/Tensile/TensileCreateLib/ParseArguments.py
@@ -265,7 +265,7 @@ def parseArguments(input: Optional[List[str]] = None) -> Dict[str, Any]:
         "--ignore-asm-cap-cache",
         dest="IgnoreAsmCapCache",
         action="store_true",
-        default=False,
+        default=True,
         help="Ignore ASM capabilities cache and derive the capabilities at runtime.",
     )
     parser.add_argument(
diff --git a/docs/src/cli-reference/tensile-create-library-cli.rst b/docs/src/cli-reference/tensile-create-library-cli.rst
index 3fd61f6f..6caad8c8 100644
--- a/docs/src/cli-reference/tensile-create-library-cli.rst
+++ b/docs/src/cli-reference/tensile-create-library-cli.rst
@@ -54,9 +54,9 @@ Here is the list of optional arguments for invoking the ``TensileCreateLibrary``
     - Architectures to generate a library for. When specifying multiple options, use quoted and semicolon-delimited
       architectures such as \-\-architecture='gfx908;gfx1012'.
       Supported architectures include: all; gfx000; gfx803; gfx900; gfx900:xnack-; gfx906; gfx906:xnack+; gfx906:xnack-; gfx908; gfx908:xnack+;
-      gfx908:xnack-; gfx90a; gfx90a:xnack+; gfx90a:xnack-; gfx940; gfx940:xnack+; gfx940:xnack-; gfx941; gfx941:xnack+;
+      gfx908:xnack-; gfx90a; gfx90a:xnack+; gfx90a:xnack-; gfx90c; gfx940; gfx940:xnack+; gfx940:xnack-; gfx941; gfx941:xnack+;
       gfx941:xnack-; gfx942; gfx942:xnack+; gfx942:xnack-; gfx1010; gfx1011; gfx1012; gfx1030; gfx1031; gfx1032; gfx1034; gfx1035;
-      gfx1100; gfx1101; gfx1102; gfx1150; gfx1151.
+      gfx1100; gfx1101; gfx1102; gfx1103; gfx1150; gfx1151.
 
   * - \-\-build-client
     - Builds Tensile client executable that is used for stand alone benchmarking. This option is set by default.
diff --git a/pytest.ini b/pytest.ini
index c0ea3fcd..fe17c9da 100644
--- a/pytest.ini
+++ b/pytest.ini
@@ -92,6 +92,7 @@ markers =
  xfail-gfx906:  architecture
  xfail-gfx908:  architecture
  xfail-gfx90a:  architecture
+ xfail-gfx90c:  architecture
  xfail-gfx940:  architecture
  xfail-gfx941:  architecture
  xfail-gfx942:  architecture
@@ -107,6 +108,7 @@ markers =
  xfail-gfx1100: architecture
  xfail-gfx1101: architecture
  xfail-gfx1102: architecture
+ xfail-gfx1103: architecture
  xfail-gfx1150: architecture
  xfail-gfx1151: architecture
  skip-gfx000:  architecture
@@ -114,6 +116,7 @@ markers =
  skip-gfx906:  architecture
  skip-gfx908:  architecture
  skip-gfx90a:  architecture
+ skip-gfx90c:  architecture
  skip-gfx940:  architecture
  skip-gfx941:  architecture
  skip-gfx942:  architecture
@@ -129,5 +132,6 @@ markers =
  skip-gfx1100: architecture
  skip-gfx1101: architecture
  skip-gfx1102: architecture
+ skip-gfx1103: architecture
  skip-gfx1150: architecture 
  skip-gfx1151: architecture
