diff --git a/src/compiler/CMakeLists.txt b/src/compiler/CMakeLists.txt
index ab80a46..31979a6 100644
--- a/src/compiler/CMakeLists.txt
+++ b/src/compiler/CMakeLists.txt
@@ -87,8 +87,8 @@ if(ROCM_FOUND)
                 atmi_cplugin
                 ${ROCM_LIBRARIES}
                 -L${ROCM_LIBRARIES_DIR}/../hsa/lib -L${ROCM_LIBRARIES_DIR}
-                -Wl,--enable-new-dtags
                 -Wl,-rpath,\$ORIGIN
+                -Wl,-rpath,\$ORIGIN/../lib
                 -Wl,-rpath,\$ORIGIN/../../hsa/lib -Wl,-rpath,\$ORIGIN/../../lib
             )
        else()
diff --git a/src/runtime/core/CMakeLists.txt b/src/runtime/core/CMakeLists.txt
index 29abffb..8f4abfb 100644
--- a/src/runtime/core/CMakeLists.txt
+++ b/src/runtime/core/CMakeLists.txt
@@ -111,8 +111,8 @@ target_link_libraries(
   ${ROCM_LIBRARIES}
   -L${ROCM_LIBRARIES_DIR}/../hsa/lib
   -L${ROCM_LIBRARIES_DIR}
-  -Wl,--enable-new-dtags
   -Wl,-rpath,\$ORIGIN
+  -Wl,-rpath,\$ORIGIN/../lib
   -Wl,-rpath,\$ORIGIN/../../hsa/lib
   -Wl,-rpath,\$ORIGIN/../../lib
 )
