diff --git a/src/main.cpp b/src/main.cpp
index abfc14e..027bb21 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -210,9 +210,8 @@ void run()
     else if (selection == Benchmark::Triad)
     {
       std::cout << "Running triad " << num_times << " times" << std::endl;
-      std::cout << "Number of elements: " << ARRAY_SIZE << std::endl;
     }
-
+    std::cout << "Number of elements: " << ARRAY_SIZE << std::endl;
 
     if (sizeof(T) == sizeof(float))
       std::cout << "Precision: float" << std::endl;
diff --git a/src/omp/OMPStream.h b/src/omp/OMPStream.h
index 5a5622f..446a3a9 100644
--- a/src/omp/OMPStream.h
+++ b/src/omp/OMPStream.h
@@ -7,6 +7,10 @@
 
 #pragma once
 
+#ifdef OMP_USM
+#pragma omp requires unified_shared_memory
+#endif
+
 #include <iostream>
 #include <stdexcept>
 
