diff --git a/opencl/tools/clinfo/CMakeLists.txt b/opencl/tools/clinfo/CMakeLists.txt
index 0af9f1eab..b2bcb7904 100644
--- a/opencl/tools/clinfo/CMakeLists.txt
+++ b/opencl/tools/clinfo/CMakeLists.txt
@@ -3,13 +3,18 @@ add_executable(clinfo clinfo.cpp)
 target_compile_definitions(clinfo PRIVATE CL_TARGET_OPENCL_VERSION=220 HAVE_CL2_HPP)
 
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../../cmake")
-find_package(AMD_ICD)
 
 #todo: to be updated to use header files from other repos
 target_include_directories(clinfo PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/../../khronos/headers/opencl2.2")
 
-find_library(AMD_ICD_LIBRARY OpenCL HINTS "${AMD_ICD_LIBRARY_DIR}")
-target_link_libraries(clinfo PRIVATE ${AMD_ICD_LIBRARY})
+if(BUILD_ICD)
+  add_dependencies(clinfo OpenCL)
+  target_link_libraries(clinfo PRIVATE OpenCL)
+else()
+  find_package(AMD_ICD)
+  find_library(AMD_ICD_LIBRARY OpenCL HINTS "${AMD_ICD_LIBRARY_DIR}")
+  target_link_libraries(clinfo PRIVATE ${AMD_ICD_LIBRARY})
+endif()
 
 INSTALL(TARGETS clinfo
   RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
diff --git a/hipamd/CMakeLists.txt b/hipamd/CMakeLists.txt
index 98f37b9d1..0546d7783 100755
--- a/hipamd/CMakeLists.txt
+++ b/hipamd/CMakeLists.txt
@@ -415,8 +415,8 @@ if (NOT ${HIPCC_BIN_DIR} STREQUAL "")
   if(EXISTS ${HIPCC_BIN_DIR})
     install(PROGRAMS ${HIPCC_BIN_DIR}/${HIPCC_EXECUTABLE} DESTINATION bin)
     install(PROGRAMS ${HIPCC_BIN_DIR}/${HIPCONFIG_EXECUTABLE} DESTINATION bin)
-    install(PROGRAMS ${HIPCC_BIN_DIR}/hipcc.pl DESTINATION bin)
-    install(PROGRAMS ${HIPCC_BIN_DIR}/hipconfig.pl DESTINATION bin)
+#    install(PROGRAMS ${HIPCC_BIN_DIR}/hipcc.pl DESTINATION bin)
+#    install(PROGRAMS ${HIPCC_BIN_DIR}/hipconfig.pl DESTINATION bin)
     install(PROGRAMS ${HIPCC_BIN_DIR}/hipvars.pm DESTINATION bin)
 
     if(NOT UNIX)
