diff --git a/lib/comgr/src/comgr-disassembly.cpp b/lib/comgr/src/comgr-disassembly.cpp
index 57cf639..5008c29 100644
--- a/lib/comgr/src/comgr-disassembly.cpp
+++ b/lib/comgr/src/comgr-disassembly.cpp
@@ -144,7 +144,7 @@ amd_comgr_status_t DisassemblyInfo::disassembleInstruction(uint64_t Address,
 
   std::string InstStr;
   raw_string_ostream InstStream(InstStr);
-  IP->printInst(&Inst, Address, AnnotationsStream.str(), *STI, InstStream);
+  IP->printInst(&Inst, InstStream, AnnotationsStream.str(), *STI);
 
   PrintInstruction(InstStream.str().c_str(), UserData);
 
diff --git a/lib/comgr/src/comgr-objdump.cpp b/lib/comgr/src/comgr-objdump.cpp
index 8462b2c..f5a7ff7 100644
--- a/lib/comgr/src/comgr-objdump.cpp
+++ b/lib/comgr/src/comgr-objdump.cpp
@@ -560,7 +560,7 @@ public:
       dumpBytes(Bytes, OS);
     }
     if (MI)
-      IP.printInst(MI, Address.Address, "", STI, OS);
+      IP.printInst(MI, OS, "", STI);
     else
       OS << " <unknown>";
   }
@@ -593,7 +593,7 @@ public:
     std::string Buffer;
     {
       raw_string_ostream TempStream(Buffer);
-      IP.printInst(MI, Address.Address, "", STI, TempStream);
+      IP.printInst(MI, TempStream, "", STI);
     }
     StringRef Contents(Buffer);
     // Split off bundle attributes
@@ -645,7 +645,7 @@ public:
     SmallString<40> InstStr;
     raw_svector_ostream IS(InstStr);
 
-    IP.printInst(MI, Address.Address, "", STI, IS);
+    IP.printInst(MI, IS, "", STI);
 
     OS << left_justify(IS.str(), 60)
        << format("// %012" PRIX64 ": ", Address.Address);
@@ -677,7 +677,7 @@ public:
       dumpBytes(Bytes, OS);
     }
     if (MI)
-      IP.printInst(MI, Address.Address, "", STI, OS);
+      IP.printInst(MI, OS, "", STI);
     else
       OS << " <unknown>";
   }
