diff --git a/Modules/Basics/DataManagement/ArrayOperations/Clear_Command.f90 b/Modules/Basics/DataManagement/ArrayOperations/Clear_Command.f90
index eba18541..6d430130 100644
--- a/Modules/Basics/DataManagement/ArrayOperations/Clear_Command.f90
+++ b/Modules/Basics/DataManagement/ArrayOperations/Clear_Command.f90
@@ -49,9 +49,9 @@ contains
     integer ( KDI ), dimension ( : ), intent ( out ) :: &
       A
 
-    !$OMP parallel workshare
+!    !$OMP parallel workshare
     A = 0_KDI
-    !$OMP end parallel workshare
+!    !$OMP end parallel workshare
 
   end subroutine ClearInteger_1D
 
@@ -61,9 +61,9 @@ contains
     integer ( KDI ), dimension ( :, : ), intent ( out ) :: &
       A
 
-    !$OMP parallel workshare
+!    !$OMP parallel workshare
     A = 0_KDI
-    !$OMP end parallel workshare
+!    !$OMP end parallel workshare
 
   end subroutine ClearInteger_2D
 
@@ -73,9 +73,9 @@ contains
     integer ( KDI ), dimension ( :, :, : ), intent ( out ) :: &
       A
 
-    !$OMP parallel workshare
+!    !$OMP parallel workshare
     A = 0_KDI
-    !$OMP end parallel workshare
+!    !$OMP end parallel workshare
 
   end subroutine ClearInteger_3D
 
diff --git a/Modules/Basics/DataManagement/ArrayOperations/Copy_Command.f90 b/Modules/Basics/DataManagement/ArrayOperations/Copy_Command.f90
index e2f7ae74..0ea9ace6 100644
--- a/Modules/Basics/DataManagement/ArrayOperations/Copy_Command.f90
+++ b/Modules/Basics/DataManagement/ArrayOperations/Copy_Command.f90
@@ -63,9 +63,9 @@ contains
     integer ( KDI ), dimension ( : ), intent ( out ) :: &
       B
 
-    !$OMP parallel workshare
+!    !$OMP parallel workshare
     B = A
-    !$OMP end parallel workshare
+!    !$OMP end parallel workshare
 
   end subroutine CopyInteger_1D
   
@@ -183,9 +183,9 @@ contains
     integer ( KBI ), dimension ( : ), intent ( out ) :: &
       B
 
-    !$OMP parallel workshare
+!    !$OMP parallel workshare
     B = A
-    !$OMP end parallel workshare
+!    !$OMP end parallel workshare
 
   end subroutine CopyBigInteger_1D
   
diff --git a/Modules/Basics/Display/CONSOLE_Singleton.f90 b/Modules/Basics/Display/CONSOLE_Singleton.f90
index 2f8cd18a..aa5b4ac0 100644
--- a/Modules/Basics/Display/CONSOLE_Singleton.f90
+++ b/Modules/Basics/Display/CONSOLE_Singleton.f90
@@ -18,7 +18,7 @@ module CONSOLE_Singleton
     logical ( KDL ) :: &
       Muted = .false.
     procedure ( AbortInterface ), nopass, pointer :: &
-      Abort => null ( )
+      Abort
   contains
     procedure, public, nopass :: &
       Initialize
@@ -180,7 +180,7 @@ contains
   end subroutine Unmute
   
 
-  elemental subroutine Finalize ( C )
+  impure elemental subroutine Finalize ( C )
 
     type ( ConsoleSingleton ), intent ( inout ) :: &
       C
diff --git a/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90 b/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90
index a2d813c..335e77b 100644
--- a/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90
+++ b/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90
@@ -104,8 +104,14 @@ contains
     VZ = K ( 3 ) / ( dot_product ( K, K ) * Period )

     end associate !-- K

+#ifdef ENABLE_OMP_OFFLOAD
+      call PF % UpdateDevice ( )
+#endif
     call PF % ComputeAuxiliary ( PF % Value )
     call PF % ComputeConserved ( PF % Value )
+#ifdef ENABLE_OMP_OFFLOAD
+      call PF % UpdateHost ( )
+#endif

     !-- FIXME: Implicit do-loop array needed to workaround Cray compiler
     !          crashes for elemental function argument of array and scalar
diff --git a/Modules/Basics/Devices/Device_HIP.c b/Modules/Basics/Devices/Device_HIP.c
index cd2c80a4..1aaed77d 100644
--- a/Modules/Basics/Devices/Device_HIP.c
+++ b/Modules/Basics/Devices/Device_HIP.c
@@ -1,3 +1,4 @@
+#ifndef USE_OPENMP_RUNTIME
 #include <stdlib.h>
 #include <stdio.h>

@@ -86,3 +87,4 @@ int DeviceMemGetInfo_Device ( size_t * Free, size_t * Total )
   return -1;
   #endif
   }
+#endif
diff --git a/Modules/Basics/Devices/Device_OMP.c b/Modules/Basics/Devices/Device_OMP.c
index abccbb8d..0aca89fb 100644
--- a/Modules/Basics/Devices/Device_OMP.c
+++ b/Modules/Basics/Devices/Device_OMP.c
@@ -220,3 +220,54 @@ bool OffloadEnabled ( )
   return false;
   #endif
   }
+
+#ifdef USE_OPENMP_RUNTIME
+int SetDevice ( int iDevice )
+  {
+  #ifdef ENABLE_OMP_OFFLOAD
+  omp_set_default_device(iDevice);
+  return 0;
+  #else
+  return -1;
+  #endif
+  }
+
+
+int GetDevice ( int * iDevice )
+  {
+  #ifdef ENABLE_OMP_OFFLOAD
+  *iDevice = omp_get_default_device();
+  return 0;
+  #else
+  return -1;
+  #endif
+  }
+
+void * AllocateHostDouble_Device ( int nValues )
+  {
+  void * Host;
+
+  #ifdef ENABLE_OMP_OFFLOAD
+  Host = omp_target_alloc( sizeof ( double ) * nValues, omp_get_initial_device());
+  #else
+  Host = malloc ( sizeof ( double ) * nValues );
+  #endif
+  return Host;
+  }
+
+void FreeHost_Device ( void * Host )
+  {
+  #ifdef ENABLE_OMP_OFFLOAD
+  omp_target_free(Host, omp_get_initial_device());
+  #else
+  free ( Host );
+  #endif
+  }
+
+int DeviceMemGetInfo_Device ( size_t * Free, size_t * Total )
+  {
+  Free  = 0;
+  Total = 0;
+  return -1;
+  }
+#endif

