diff --git a/Build/Preprocessor b/Build/Preprocessor
index 5fc7364..4ae10ae 100644
--- a/Build/Preprocessor
+++ b/Build/Preprocessor
@@ -9,7 +9,7 @@
 #define OMP_TARGET_DISTRIBUTE_SCHEDULE dist_schedule ( static, 1 )
 #endif
 
-#if defined(POWER_GCC) || defined(Cray_GCC) || defined(Cray_GCC_UMS) || defined (Cray_CCE) || defined (Cray_AMD)
+#if defined(POWER_GCC) || defined(Cray_GCC) || defined(Cray_GCC_UMS) || defined (Cray_CCE) || defined (Cray_AMD) || defined(ROCm) || defined(ROCmFlangNew)
 #define OMP_SCHEDULE_TARGET auto
 #define OMP_SCHEDULE_HOST runtime
 #define OMP_TARGET_DISTRIBUTE_SCHEDULE dist_schedule ( auto )
diff --git a/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90 b/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90
index a2d813c..580d9ab 100644
--- a/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90
+++ b/Programs/Examples/Basics/FluidDynamics/PlaneWaveAdvection_Template.f90
@@ -103,8 +103,14 @@ contains

     end associate !-- K

+#ifdef ENABLE_OMP_OFFLOAD
+      call PF % UpdateDevice ( )
+#endif
     call PF % ComputeAuxiliary ( PF % Value )
     call PF % ComputeConserved ( PF % Value )
+#ifdef ENABLE_OMP_OFFLOAD
+      call PF % UpdateHost ( )
+#endif

     !-- FIXME: Implicit do-loop array needed to workaround Cray compiler
     !          crashes for elemental function argument of array and scalar
