diff --git a/include/hip/hcc_detail/device_functions.h b/include/hip/hcc_detail/device_functions.h
index e2ffacca..bd6a86f4 100644
--- a/include/hip/hcc_detail/device_functions.h
+++ b/include/hip/hcc_detail/device_functions.h
@@ -964,15 +964,7 @@ __device__ static inline float __ull2float_rn(unsigned long long int x) { return
 __device__ static inline float __ull2float_ru(unsigned long long int x) { return (float)x; }
 __device__ static inline float __ull2float_rz(unsigned long long int x) { return (float)x; }
 
-#if defined(__HCC__)
-#define __HCC_OR_HIP_CLANG__ 1
-#elif defined(__clang__) && defined(__HIP__)
-#define __HCC_OR_HIP_CLANG__ 1
-#else
-#define __HCC_OR_HIP_CLANG__ 0
-#endif
-
-#ifdef __HCC_OR_HIP_CLANG__
+#if defined(__HCC__) || ( defined(__clang__) && defined(__HIP__))
 
 // Clock functions
 __device__ long long int __clock64();
@@ -1359,6 +1351,7 @@ unsigned __smid(void)
 
 #endif //defined(__clang__) && defined(__HIP__)
 
+#if defined(__HCC__) || ( defined(__clang__) && defined(__HIP__))
 
 // loop unrolling
 static inline __device__ void* __hip_hc_memcpy(void* dst, const void* src, size_t size) {
@@ -1420,3 +1413,5 @@ static inline __device__ void* memset(void* ptr, int val, size_t size) {
 }
 
 #endif
+
+#endif
diff --git a/include/hip/hcc_detail/hip_runtime.h b/include/hip/hcc_detail/hip_runtime.h
index 0a173bb4..ee617b43 100644
--- a/include/hip/hcc_detail/hip_runtime.h
+++ b/include/hip/hcc_detail/hip_runtime.h
@@ -312,6 +312,8 @@ static constexpr Coordinates<hip_impl::WorkitemId> threadIdx{};
 #define hipGridDim_z (hc_get_num_groups(2))
 
 #endif // defined __HCC__
+
+#ifndef __OPENMP_AMDGCN__
 #if __HCC_OR_HIP_CLANG__
 #if !__CLANG_HIP_RUNTIME_WRAPPER_INCLUDED__
 #if __HIP_ENABLE_DEVICE_MALLOC__
@@ -325,6 +327,7 @@ static inline __device__ void* free(void* ptr) { __builtin_trap(); return nullpt
 #endif
 #endif // !__CLANG_HIP_RUNTIME_WRAPPER_INCLUDED__
 #endif //__HCC_OR_HIP_CLANG__
+#endif // not openmp
 
 #ifdef __HCC__
 
diff --git a/include/hip/hcc_detail/math_fwd.h b/include/hip/hcc_detail/math_fwd.h
index 34b38aa7..c197af89 100644
--- a/include/hip/hcc_detail/math_fwd.h
+++ b/include/hip/hcc_detail/math_fwd.h
@@ -364,16 +364,16 @@ __attribute__((const))
 float __ocml_div_rtz_f32(float, float);
 __device__
 __attribute__((const))
-float __ocml_sqrt_rte_f32(float, float);
+float __ocml_sqrt_rte_f32(float);
 __device__
 __attribute__((const))
-float __ocml_sqrt_rtn_f32(float, float);
+float __ocml_sqrt_rtn_f32(float);
 __device__
 __attribute__((const))
-float __ocml_sqrt_rtp_f32(float, float);
+float __ocml_sqrt_rtp_f32(float);
 __device__
 __attribute__((const))
-float __ocml_sqrt_rtz_f32(float, float);
+float __ocml_sqrt_rtz_f32(float);
 __device__
 __attribute__((const))
 float __ocml_fma_rte_f32(float, float, float);
@@ -675,16 +675,16 @@ __attribute__((const))
 double __ocml_div_rtz_f64(double, double);
 __device__
 __attribute__((const))
-double __ocml_sqrt_rte_f64(double, double);
+double __ocml_sqrt_rte_f64(double);
 __device__
 __attribute__((const))
-double __ocml_sqrt_rtn_f64(double, double);
+double __ocml_sqrt_rtn_f64(double);
 __device__
 __attribute__((const))
-double __ocml_sqrt_rtp_f64(double, double);
+double __ocml_sqrt_rtp_f64(double);
 __device__
 __attribute__((const))
-double __ocml_sqrt_rtz_f64(double, double);
+double __ocml_sqrt_rtz_f64(double);
 __device__
 __attribute__((const))
 double __ocml_fma_rte_f64(double, double, double);
