diff --git a/include/hip/hcc_detail/hip_runtime.h b/include/hip/hcc_detail/hip_runtime.h
index c4ef1d82..bab217f6 100644
--- a/include/hip/hcc_detail/hip_runtime.h
+++ b/include/hip/hcc_detail/hip_runtime.h
@@ -66,7 +66,6 @@ THE SOFTWARE.
 // Remainder of this file only compiles with HCC
 #if defined __HCC__
 #include "grid_launch.h"
-#include "hc_printf.hpp"
 // TODO-HCC-GL - change this to typedef.
 // typedef grid_launch_parm hipLaunchParm ;
 
@@ -257,7 +256,7 @@ template <typename... All>
 static inline __device__ void printf(const char* format, All... all) {
     hc::printf(format, all...);
 }
-#elif defined(__HCC_ACCELERATOR__) || __HIP__
+#elif defined(__HCC_ACCELERATOR__)
 template <typename... All>
 static inline __device__ void printf(const char* format, All... all) {}
 #endif
