diff --git a/flang-rt/lib/runtime/io-api-minimal.cpp b/flang-rt/lib/runtime/io-api-minimal.cpp
index fdf7183ed517..f84b62d63baa 100644
--- a/flang-rt/lib/runtime/io-api-minimal.cpp
+++ b/flang-rt/lib/runtime/io-api-minimal.cpp
@@ -19,7 +19,9 @@
 #include "flang/Runtime/io-api.h"
 
 namespace Fortran::runtime::io {
+#ifdef RT_OFFLOAD_IO
 RT_EXT_API_GROUP_BEGIN
+#endif
 
 Cookie IODEF(BeginExternalListOutput)(
     ExternalUnit unitNumber, const char *sourceFile, int sourceLine) {
@@ -161,4 +163,6 @@ void std::__libcpp_verbose_abort(char const *format, ...) noexcept(
 }
 #endif
 
+#ifdef RT_OFFLOAD_IO
 RT_EXT_API_GROUP_END
+#endif
diff --git a/flang-rt/lib/runtime/io-api.cpp b/flang-rt/lib/runtime/io-api.cpp
index 0355734c67fc..ff846bbc10cb 100644
--- a/flang-rt/lib/runtime/io-api.cpp
+++ b/flang-rt/lib/runtime/io-api.cpp
@@ -31,7 +31,9 @@
 #include <memory>
 
 namespace Fortran::runtime::io {
+#ifdef RT_OFFLOAD_IO
 RT_EXT_API_GROUP_BEGIN
+#endif
 
 RT_API_ATTRS const char *InquiryKeywordHashDecode(
     char *buffer, std::size_t n, InquiryKeywordHash hash) {
@@ -1315,5 +1317,7 @@ enum Iostat IODEF(CheckUnitNumberInRange128)(common::int128_t unit,
 }
 #endif
 
+#ifdef RT_OFFLOAD_IO
 RT_EXT_API_GROUP_END
+#endif
 } // namespace Fortran::runtime::io
diff --git a/flang-rt/lib/runtime/terminator.cpp b/flang-rt/lib/runtime/terminator.cpp
index 8a57ba06b130..719441062fa1 100644
--- a/flang-rt/lib/runtime/terminator.cpp
+++ b/flang-rt/lib/runtime/terminator.cpp
@@ -45,7 +45,8 @@ RT_API_ATTRS void Terminator::CrashHeader() const {
 #if defined(RT_DEVICE_COMPILATION)
   std::printf("\nfatal Fortran runtime error");
   if (sourceFileName_) {
-    std::printf("(%s", sourceFileName_);
+    // commenting out temporarily to avoid ICE seen with amd-staging
+    // std::printf("(%s", sourceFileName_);
     if (sourceLine_) {
       std::printf(":%d", sourceLine_);
     }
diff --git a/flang/include/flang/Runtime/freestanding-tools.h b/flang/include/flang/Runtime/freestanding-tools.h
index 8a4fbcff31b4..de59f7b13ab1 100644
--- a/flang/include/flang/Runtime/freestanding-tools.h
+++ b/flang/include/flang/Runtime/freestanding-tools.h
@@ -110,12 +110,12 @@ using std::fill_n;
 
 #if STD_MEMSET_USE_BUILTIN
 static inline RT_API_ATTRS void memset(
-    void *dest, uint8_t value, std::size_t count) {
+    void *dest, unsigned char value, std::size_t count) {
   __builtin_memset(dest, value, count);
 }
 #elif STD_MEMSET_UNSUPPORTED
 static inline RT_API_ATTRS void memset(
-    void *dest, uint8_t value, std::size_t count) {
+    void *dest, unsigned char value, std::size_t count) {
   char *to{reinterpret_cast<char *>(dest)};
   while (count--) {
     *to++ = value;
