diff --git a/Makefile.defs b/Makefile.defs
index 941704e..0846f05 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -11,9 +11,9 @@ endif
 ifeq ($(GLOMPRTL),)
   $(error GLOMPRTL is not set)
 endif
-ifeq ($(CUDAVERSION),)
-  $(error CUDAVERSION is not set)
-endif
+#ifeq ($(CUDAVERSION),)
+#  $(error CUDAVERSION is not set)
+#endif
 ifeq ($(LLVMBIN),)
   $(error LLVMBIN is not set)
 endif
@@ -87,8 +87,11 @@ else ifeq ($(machine),x86_64)
 
   cmdx += -std=c++11 -v
   
-  cmdc += -L$(HOSTRTL) -L$(TARGETRTL) -target x86_64-pc-linux-gnu -fopenmp=libomp -fopenmp-targets=nvptx64-nvidia-cuda
-  cmdx += -L$(HOSTRTL) -L$(TARGETRTL) -target x86_64-pc-linux-gnu -fopenmp=libomp -fopenmp-targets=nvptx64-nvidia-cuda
+  #cmdc += -L$(HOSTRTL) -L$(TARGETRTL) -target x86_64-pc-linux-gnu -fopenmp=libomp -fopenmp-targets=nvptx64-nvidia-cuda
+  #cmdx += -L$(HOSTRTL) -L$(TARGETRTL) -target x86_64-pc-linux-gnu -fopenmp=libomp -fopenmp-targets=nvptx64-nvidia-cuda
+
+  cmdc += -L$(HOSTRTL) -L$(TARGETRTL) $(TARGET)
+  cmdx += -L$(HOSTRTL) -L$(TARGETRTL) $(TARGET)
 
   extra_ld_lib_path := /home/compteam/opt/gcc/lib64
 else
@@ -96,6 +99,7 @@ else
 endif
 
 main_folder := $(shell pwd)
+program := $(shell basename $(main_folder))
 results_folder := $(main_folder)/results
 error_log := $(results_folder)/error.log
 compile_log := $(results_folder)/compile.log
