diff --git a/include/RAJA/policy/atomic_auto.hpp b/include/RAJA/policy/atomic_auto.hpp
index 34d7f50a..b323adae 100644
--- a/include/RAJA/policy/atomic_auto.hpp
+++ b/include/RAJA/policy/atomic_auto.hpp
@@ -36,7 +36,8 @@
  * Finally, we fallback on the seq_atomic, which performs non-atomic operations
  * because we assume there is no thread safety issues (no parallel model)
  */
-#if defined(__CUDA_ARCH__)
+
+#if defined(RAJA_ENABLE_CUDA) && defined(__CUDA_ARCH__)
 #define RAJA_AUTO_ATOMIC \
   RAJA::cuda_atomic {}
 #elif defined(RAJA_ENABLE_OPENMP)
