diff --git a/lib/comgr/CMakeLists.txt b/lib/comgr/CMakeLists.txt
index 412bd2e..32f1734 100644
--- a/lib/comgr/CMakeLists.txt
+++ b/lib/comgr/CMakeLists.txt
@@ -17,7 +17,10 @@ if (ROCM_CCACHE_BUILD)
   endif() # if (CCACHE_PROGRAM)
 endif() # if (ROCM_CCACHE_BUILD)
 
+if(NOT AMD_COMGR_BUILD_NO_ROCM)
 find_package(ROCM PATHS "/opt/rocm")
+endif()
+
 if (ROCM_FOUND)
   include(ROCMSetupVersion)
   rocm_setup_version(VERSION "${amd_comgr_VERSION}")
diff --git a/lib/comgr/cmake/DeviceLibs.cmake b/lib/comgr/cmake/DeviceLibs.cmake
index 27e9546..29cc2b9 100644
--- a/lib/comgr/cmake/DeviceLibs.cmake
+++ b/lib/comgr/cmake/DeviceLibs.cmake
@@ -115,10 +115,10 @@ endforeach()
 file(APPEND ${GEN_LIBRARY_INC_FILE} "\n#include \"llvm/ADT/ArrayRef.h\"\n")
 file(APPEND ${GEN_LIBRARY_INC_FILE}
   "llvm::ArrayRef<std::tuple<llvm::StringRef, llvm::StringRef>> COMGR::getDeviceLibraries() { \
-   static std::tuple<llvm::StringRef, llvm::StringRef> DeviceLibs[] = {")
+llvm::ArrayRef<std::tuple<llvm::StringRef, llvm::StringRef>> DeviceLibs = {")
 foreach(AMDGCN_LIB_TARGET ${AMD_DEVICE_LIBS_TARGETS})
   file(APPEND ${GEN_LIBRARY_INC_FILE}
-    "{\"${AMDGCN_LIB_TARGET}.bc\", llvm::StringRef(reinterpret_cast<const char *>(${AMDGCN_LIB_TARGET}_lib), ${AMDGCN_LIB_TARGET}_lib_size)},")
+    " std::tuple<llvm::StringRef, llvm::StringRef> (\"${AMDGCN_LIB_TARGET}.bc\", llvm::StringRef(reinterpret_cast<const char *>(${AMDGCN_LIB_TARGET}_lib), ${AMDGCN_LIB_TARGET}_lib_size)),")
 endforeach()
 file(APPEND ${GEN_LIBRARY_INC_FILE}
   "}; \
