diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6d142bc..8309bd5 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -147,6 +147,10 @@ if((${CMAKE_HOST_SYSTEM_PROCESSOR} STREQUAL "x86_64") OR (${CMAKE_HOST_SYSTEM_PR
   set(ROCMINFO_CXX_FLAGS ${ROCMINFO_CXX_FLAGS} -msse -msse2)
 endif()
 
+if(${CMAKE_HOST_SYSTEM_PROCESSOR} STREQUAL "ppc64le")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mcpu=power9")
+endif()
+
 #
 # Add compiler flags to include symbol information for debug builds
 #
