diff --git a/lib/comgr/CMakeLists.txt b/lib/comgr/CMakeLists.txt
index 1794a07..8e7797e 100644
--- a/lib/comgr/CMakeLists.txt
+++ b/lib/comgr/CMakeLists.txt
@@ -16,7 +16,9 @@ endif() # if (ROCM_CCACHE_BUILD)
 
 project(amd_comgr VERSION "1.6.0" LANGUAGES C CXX)
 
+if(NOT AMD_COMGR_BUILD_NO_ROCM)
 find_package(ROCM PATHS "/opt/rocm")
+endif()
 if (ROCM_FOUND)
   include(ROCMSetupVersion)
   rocm_setup_version(VERSION "${amd_comgr_VERSION}")
@@ -237,8 +239,10 @@ if (NOT UNIX)
     PRIVATE version)
 endif()
 
+if(NOT AMD_COMGR_BUILD_NO_ROCM)
 enable_testing()
 add_subdirectory(test)
+endif()
 
 # Add packaging directives for amd_comgr
 set(CPACK_PACKAGE_NAME comgr)
diff --git a/lib/comgr/src/comgr-compiler.cpp b/lib/comgr/src/comgr-compiler.cpp
index 204ce73..af901b5 100644
--- a/lib/comgr/src/comgr-compiler.cpp
+++ b/lib/comgr/src/comgr-compiler.cpp
@@ -561,7 +561,7 @@ static amd_comgr_status_t linkWithLLD(llvm::ArrayRef<const char *> Args,
   ArgStringList LLDArgs(llvm::iterator_range<ArrayRef<const char *>::iterator>(
       Args.begin(), Args.end()));
   LLDArgs.insert(LLDArgs.begin(), "lld");
-  LLDArgs.push_back("--threads=1");
+  LLDArgs.push_back("--no-threads");
   ArrayRef<const char *> ArgRefs = llvm::makeArrayRef(LLDArgs);
   static std::mutex MScreen;
   MScreen.lock();
diff --git a/lib/comgr/test/CMakeLists.txt b/lib/comgr/test/CMakeLists.txt
index 2f96329..d423093 100644
--- a/lib/comgr/test/CMakeLists.txt
+++ b/lib/comgr/test/CMakeLists.txt
@@ -5,7 +5,7 @@ set(TEST_INPUT_BINARIES)
 macro(add_test_input_binary name input output)
   add_custom_command(
     OUTPUT "${output}"
-    COMMAND "$<TARGET_FILE:clang>" --target=amdgcn-amd-amdhsa -mcpu=gfx803 -nogpulib
+    COMMAND "$<TARGET_FILE:clang>" --target=amdgcn-amd-amdhsa -mcpu=gfx803
     ${ARGN} "${CMAKE_CURRENT_SOURCE_DIR}/${input}" -o "${output}"
     VERBATIM
     DEPENDS clang "${input}")
