diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 0eb5dea..1ec0edb 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -260,7 +260,7 @@ endif()
 
 ## Link dependencies.
 target_link_libraries ( ${CORE_RUNTIME_TARGET} PRIVATE hsakmt::hsakmt )
-target_link_libraries ( ${CORE_RUNTIME_TARGET} PRIVATE elf::elf dl pthread rt )
+target_link_libraries ( ${CORE_RUNTIME_TARGET} PRIVATE elf::elf dl pthread rt numa)
 
 ## Set the VERSION and SOVERSION values
 set_property ( TARGET ${CORE_RUNTIME_TARGET} PROPERTY VERSION "${SO_VERSION_STRING}" )
@@ -281,7 +281,7 @@ if( NOT ${BUILD_SHARED_LIBS} )
 
   ## Add external link requirements.
   target_link_libraries ( ${CORE_RUNTIME_NAME} INTERFACE hsakmt::hsakmt )
-  target_link_libraries ( ${CORE_RUNTIME_NAME} INTERFACE elf::elf dl pthread rt )
+  target_link_libraries ( ${CORE_RUNTIME_NAME} INTERFACE elf::elf dl pthread rt numa)
 
   install ( TARGETS ${CORE_RUNTIME_NAME} EXPORT ${CORE_RUNTIME_NAME}Targets )
 endif()
diff --git a/src/libamdhsacode/amd_elf_image.cpp b/src/libamdhsacode/amd_elf_image.cpp
index bff3f46..ebe707a 100644
--- a/src/libamdhsacode/amd_elf_image.cpp
+++ b/src/libamdhsacode/amd_elf_image.cpp
@@ -1270,7 +1270,12 @@ namespace elf {
 
     GElfImage::~GElfImage()
     {
-      elf_end(e);
+      // elf_end fails with an invalid free on libelf-0.170.so, ubuntu 18.
+      // hack way to tell we are on old libelf is the enum size.
+      if ( ELF_T_NUM == 26 )
+        free(e);
+      else
+        elf_end(e);
     }
 
     bool GElfImage::imgError()
