diff --git a/src/fmm.c b/src/fmm.c
index 37e543a..a5f0f73 100644
--- a/src/fmm.c
+++ b/src/fmm.c
@@ -1485,8 +1485,11 @@ static int bind_mem_to_numa(uint32_t node_id, void *mem,
 	node_mask = numa_bitmask_alloc(num_node);
 	if (!node_mask)
 		return -ENOMEM;
-
+#ifdef __PPC64__
+	numa_bitmask_setbit(node_mask, node_id * 8);
+#else
 	numa_bitmask_setbit(node_mask, node_id);
+#endif
 	mode |= flags.ui32.NoSubstitute ? MPOL_BIND : MPOL_PREFERRED;
 	r = mbind(mem, SizeInBytes, mode, node_mask->maskp, num_node + 1, 0);
 	numa_bitmask_free(node_mask);
