diff --git a/libc/CMakeLists.txt b/libc/CMakeLists.txt
index 46d799cbec11..e1228651d856 100644
--- a/libc/CMakeLists.txt
+++ b/libc/CMakeLists.txt
@@ -420,10 +420,9 @@ add_subdirectory(lib)
 if(LLVM_INCLUDE_TESTS)
   add_subdirectory(test)
   add_subdirectory(fuzzing)
+# add_subdirectory(benchmarks)
 endif()
 
-add_subdirectory(benchmarks)
-
 if (LIBC_INCLUDE_DOCS)
   add_subdirectory(docs)
 endif()
