diff --git a/libc/CMakeLists.txt b/libc/CMakeLists.txt
index 7cbea4d8a0a8..b656c8da4f76 100644
--- a/libc/CMakeLists.txt
+++ b/libc/CMakeLists.txt
@@ -456,10 +456,9 @@ add_subdirectory(lib)
 if(LLVM_INCLUDE_TESTS)
   add_subdirectory(test)
   add_subdirectory(fuzzing)
+# add_subdirectory(benchmarks)
 endif()
 
-add_subdirectory(benchmarks)
-
 if(LIBC_INCLUDE_DOCS)
   add_subdirectory(docs)
 endif()
