diff --git a/compiler/CMakeLists.txt b/compiler/CMakeLists.txt
index cdd56f8..37071c6 100644
--- a/compiler/CMakeLists.txt
+++ b/compiler/CMakeLists.txt
@@ -10,6 +10,20 @@ if(THEROCK_ENABLE_COMPILER)
   if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     list(APPEND _extra_llvm_cmake_args "-DLLVM_ENABLE_PEDANTIC=OFF")
   endif()
+  if (DEFINED ENV{AOMP_ENABLE_PROJECTS})
+    # AOMP typically needs these subprojects clang;lld;clang-tools-extra;flang
+    list(APPEND _extra_llvm_cmake_args "-DLLVM_ENABLE_PROJECTS='$ENV{AOMP_ENABLE_PROJECTS}'")
+  endif()
+  if (DEFINED ENV{AOMP_ENABLE_RUNTIMES})
+    # AOMP typically needs libcxx;libcxxabi;libunwind;openmp;offload;compiler-rt;flang-rt;
+    list(APPEND _extra_llvm_cmake_args "-DLLVM_ENABLE_RUNTIMES='$ENV{AOMP_ENABLE_RUNTIMES}'")
+  endif()
+  if (DEFINED ENV{AOMP_VERSION_STRING})
+    list(APPEND _extra_llvm_cmake_args "-DLLVM_VERSION_SUFFIX=_AOMP_$ENV{AOMP_VERSION_STRING}")
+  endif()
+  if (DEFINED ENV{AOMP_VERSION_MOD})
+    list(APPEND _extra_llvm_cmake_args "-DLLVM_VERSION_PATCH=$ENV{AOMP_VERSION_MOD}")
+  endif()
 
   # If the compiler is not pristine (i.e. has patches), then there will be a
   # ".amd-llvm.smrev" file present which must be used instead of the auto
@@ -68,6 +82,17 @@ if(THEROCK_ENABLE_COMPILER)
       -DCLANG_ENABLE_CLANGD=OFF
       -DCLANG_TIDY_ENABLE_STATIC_ANALYZER=OFF
 
+      -DCLANG_TOOL_CLANG_LINKER_WRAPPER_BUILD=ON
+      -DOPENMP_ENABLE_LIBOMPTARGET=ON
+      -DLIBOMPTARGET_BUILD_DEVICE_FORTRT=ON
+      -DOFFLOAD_EXTERNAL_PROJECT_UNIFIED_ROCR=ON
+      -DOPENMP_ENABLE_LIBOMPTARGET=1
+      -DLIBOMP_USE_HWLOC=ON
+      -DLIBOMPTARGET_EXTERNAL_PROJECT_HSA_PATH=${CMAKE_CURRENT_SOURCE_DIR}/../rocm-systems/projects/rocr-runtime
+      -DFLANG_RUNTIME_F128_MATH_LIB=libquadmath
+      -DROCM_DEVICE_LIBS_BITCODE_INSTALL_LOC_NEW=${CMAKE_CURRENT_BINARY_DIR}/lib/clang/${LLVM_VERSION_MAJOR}/lib/amdgcn
+      -DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=ON
+      -DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES=openmp
       ${_extra_llvm_cmake_args}
     BUILD_DEPS
       rocm-cmake
diff --git a/compiler/pre_hook_amd-llvm.cmake b/compiler/pre_hook_amd-llvm.cmake
index a01d500..da9c1eb 100644
--- a/compiler/pre_hook_amd-llvm.cmake
+++ b/compiler/pre_hook_amd-llvm.cmake
@@ -14,12 +14,28 @@ if(WIN32)
   #     Should they be supported? What depends on them?
   set(LLVM_ENABLE_LIBCXX OFF)
   set(LLVM_ENABLE_RUNTIMES "compiler-rt" CACHE STRING "Enabled runtimes" FORCE)
+  set(LLVM_ENABLE_PROJECTS "clang;lld;clang-tools-extra;flang" CACHE STRING "Enable LLVM projects" FORCE)
 else()
   set(LLVM_BUILD_LLVM_DYLIB ON)
   set(LLVM_LINK_LLVM_DYLIB ON)
   set(LLVM_ENABLE_LIBCXX ON)
-  set(LLVM_ENABLE_RUNTIMES "compiler-rt;libunwind;libcxx;libcxxabi;openmp" CACHE STRING "Enabled runtimes" FORCE)
-  # Settinng "LIBOMP_COPY_EXPORTS" to `OFF` "aids parallel builds to not interfere
+  set(LLVM_ENABLE_PROJECTS "clang;lld;clang-tools-extra;flang" CACHE STRING "Enable LLVM projects" FORCE)
+  set(LLVM_ENABLE_RUNTIMES "libcxx;libcxxabi;libunwind;openmp;offload;compiler-rt;flang-rt" CACHE STRING "Enabled runtimes" FORCE)
+
+  if("offload" IN_LIST LLVM_ENABLE_RUNTIMES)
+    set(OPENMP_ENABLE_LIBOMPTARGET ON)
+    set(LIBOMPTARGET_BUILD_DEVICE_FORTRT ON)
+    set(LIBOMPTARGET_ENABLE_DEBUG ON)
+    set(LIBOMPTARGET_NO_SANITIZER_AMDGPU ON)
+    set(LIBOMP_INSTALL_RPATH "\$ORIGIN:\$ORIGIN/../lib:\$ORIGIN/../../lib:\$ORIGIN/../../../lib")
+    # Disable building rocr-runtime as external project until libnuma-dev dependency can be satisifed
+    #set(LIBOMPTARGET_EXTERNAL_PROJECT_HSA_PATH "${THEROCK_SOURCE_DIR}/core/ROCR-Runtime")
+    #set(OFFLOAD_EXTERNAL_PROJECT_UNIFIED_ROCR ON)
+
+    # Disable ZSTD until dependency is resolved
+    set(LLVM_ENABLE_ZLIB OFF)
+  endif()
+  # Setting "LIBOMP_COPY_EXPORTS" to `OFF` "aids parallel builds to not interfere
   # with each other" as libomp and generated headers are copied into the original
   # source otherwise. Defaults to `ON`.
   set(LIBOMP_COPY_EXPORTS OFF)
@@ -27,16 +43,18 @@ endif()
 
 # Set the LLVM_ENABLE_PROJECTS variable before including LLVM's CMakeLists.txt
 set(BUILD_TESTING OFF CACHE BOOL "DISABLE BUILDING TESTS IN SUBPROJECTS" FORCE)
-set(LLVM_ENABLE_PROJECTS "clang;lld;clang-tools-extra" CACHE STRING "Enable LLVM projects" FORCE)
 set(LLVM_TARGETS_TO_BUILD "AMDGPU;X86" CACHE STRING "Enable LLVM Targets" FORCE)
+set(LLVM_RUNTIME_TARGETS "default;amdgcn-amd-amdhsa")
 
 # Packaging.
 set(PACKAGE_VENDOR "AMD" CACHE STRING "Vendor" FORCE)
 
 # Build the device-libs as part of the core compiler so that clang works by
 # default (as opposed to other components that are *users* of the compiler).
-set(LLVM_EXTERNAL_AMDDEVICE_LIBS_SOURCE_DIR "${THEROCK_SOURCE_DIR}/compiler/amd-llvm/amd/device-libs")
-set(LLVM_EXTERNAL_PROJECTS "amddevice-libs" CACHE STRING "Enable extra projects" FORCE)
+set(LLVM_EXTERNAL_PROJECTS "rocm-device-libs;SPIRV_TRANSLATOR" CACHE STRING "Enable extra projects" FORCE)
+set(LLVM_EXTERNAL_ROCM_DEVICE_LIBS_SOURCE_DIR  "${THEROCK_SOURCE_DIR}/compiler/amd-llvm/amd/device-libs")
+set(DLLVM_EXTERNAL_SPIRV_TRANSLATOR_SOURCE_DIR "${THEROCK_SOURCE_DIR}/compiler/spirv-llvm-translator")
+
 
 # TODO2: This mechanism has races in certain situations, failing to create a
 # symlink. Revisit once devicemanager code is made more robust.
@@ -47,9 +65,10 @@ set(LLVM_EXTERNAL_PROJECTS "amddevice-libs" CACHE STRING "Enable extra projects"
 # options to manage this transition but they require knowing the clange resource
 # dir. In order to avoid drift, we just fixate that too. This can all be
 # removed in a future version.
-# set(CLANG_RESOURCE_DIR "../lib/clang/${LLVM_VERSION_MAJOR}" CACHE STRING "Resource dir" FORCE)
-# set(ROCM_DEVICE_LIBS_BITCODE_INSTALL_LOC_NEW "lib/clang/${LLVM_VERSION_MAJOR}/amdgcn" CACHE STRING "New devicelibs loc" FORCE)
-# set(ROCM_DEVICE_LIBS_BITCODE_INSTALL_LOC_OLD "amdgcn" CACHE STRING "Old devicelibs loc" FORCE)
+set(CLANG_RESOURCE_DIR "../lib/clang/${LLVM_VERSION_MAJOR}" CACHE STRING "Resource dir" FORCE)
+set(ROCM_DEVICE_LIBS_BITCODE_INSTALL_LOC_NEW "lib/clang/${LLVM_VERSION_MAJOR}/lib/amdgcn" CACHE STRING "New devicelibs loc" FORCE)
+set(ROCM_DEVICE_LIBS_BITCODE_INSTALL_LOC_OLD "amdgcn" CACHE STRING "Old devicelibs loc" FORCE)
+message("= = = = = = ROCM_DEVICE_LIBS_BITCODE_INSTALL_LOC_NEW=${ROCM_DEVICE_LIBS_BITCODE_INSTALL_LOC_NEW}")
 
 # Setup the install rpath (let CMake handle build RPATH per usual):
 # * Executables and libraries can always search their adjacent lib directory
@@ -106,14 +125,14 @@ block()
     OPT
     YAML2OBJ
   )
-  if(WIN32)
+  # if(WIN32)
     # These can be provided by the "C++ Clang tools for Windows" in MSVC, but
     # we might as well build them from source ourselves.
     list(APPEND _llvm_required_tools "LLVM_AR")
     list(APPEND _llvm_required_tools "LLVM_DLLTOOL")
     list(APPEND _llvm_required_tools "LLVM_LIB")
     list(APPEND _llvm_required_tools "LLVM_RANLIB")
-  endif()
+  # endif()
   therock_set_implicit_llvm_options(LLVM "${CMAKE_CURRENT_SOURCE_DIR}/tools" "${_llvm_required_tools}")
 
   # Clang tools that are required.
@@ -123,6 +142,7 @@ block()
     CLANG_OFFLOAD_BUNDLER
     CLANG_OFFLOAD_PACKAGER
     CLANG_OFFLOAD_WRAPPER
+    CLANG_LINKER_WRAPPER
     CLANG_SHLIB
     DRIVER
   )
diff --git a/core/CMakeLists.txt b/core/CMakeLists.txt
index 3414e4b..1843ed9 100644
--- a/core/CMakeLists.txt
+++ b/core/CMakeLists.txt
@@ -108,9 +108,11 @@ if(THEROCK_ENABLE_HIP_RUNTIME)
       "-DHIP_PLATFORM=amd"
       "-DHIP_COMMON_DIR=${THEROCK_ROCM_SYSTEMS_SOURCE_DIR}/projects/hip"
       "-DCLR_BUILD_HIP=ON"
+      "-DBUILD_ICD=ON"
       # Legacy: Disable various auto-detection logic that breaks out of jail
       # and can use local machine tools.
       "-DHIPCC_BIN_DIR="
+      #      "-DHIPCC_BIN_DIR=${CMAKE_CURRENT_BINARY_DIR}/../compiler/hipcc/dist/bin"
       ${HIP_CLR_CMAKE_ARGS}
     BUILD_DEPS
       rocm-cmake
