diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7d97b713..ab16a8be 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -181,6 +181,9 @@ execute_process(
 message(STATUS "${COMPILER_EXE_NAME} HIP version:    ${hip_version_string}")
 
 ## Check for ROCm version
+if(NOT ROCM_PATH)
+  set(ROCM_PATH "${CMAKE_BINARY_DIR}/../../../dist/rocm")
+endif()
 set(EXPLICIT_ROCM_VERSION "" CACHE STRING "Explicit ROCM version to compile to (auto detect if empty)")
 if(EXPLICIT_ROCM_VERSION)
   set(rocm_version_string "${EXPLICIT_ROCM_VERSION}")
diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index 8335c72c..b7072cc3 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -83,6 +83,7 @@ if(BUILD_TESTS)
   if(RCCL_INTERFACE_COMPILE_DEFINITIONS)
     list(APPEND RCCL_COMMON_COMPILE_DEFS ${RCCL_INTERFACE_COMPILE_DEFINITIONS})
   endif()
+  target_link_libraries(rccl-UnitTests PRIVATE -L ${CMAKE_BINARY_DIR})
 
   # Collect testing framework source files
   set(TEST_SOURCE_FILES
